opal/mca/btl/smcuda/btl_smcuda_fifo.h

```diff
@@ -85,6 +85,8 @@ static void add_pending(struct mca_btl_base_endpoint_t *ep, void *data, bool res
 #define MCA_BTL_SMCUDA_FIFO_WRITE(endpoint_peer, my_smp_rank, peer_smp_rank, hdr, resend,         \
                                   retry_pending_sends, rc)                                        \
     do {                                                                                          \
+        /* memory barrier: ensure writes to the hdr have completed */                             \
+        opal_atomic_wmb();                                                                        \
```

Comment:
Is there a difference between barrier before the write vs after?

Typically I see the barrier after the actual write.

Reply:
Yes, I debated myself as well.

MCA_BTL_SMCUDA_FIFO_WRITE is called from several places.  Each of these would need an update to include a write barrier, and any new call would need to include that as well.  For this reason I felt it best to embed the barrier into the fifo_write as a "make sure hdr is committed" sort of step rather than relying on all functions which fill the header.

Reply:
The write in question here is not the integration of the header into the list but the writes related to the content of the header. This look good to me. 

Looking at the linked code it appears that the current firo_write also has a write barrier after the item integration. I can't figure out why we need that one.